src/NetAnalyzers/Core/Microsoft.NetFramework.Analyzers/DoNotUseInsecureDtdProcessingInApiDesign.cs

```diff
@@ -116,395 +104,360 @@ public void AnalyzeSymbol(SymbolAnalysisContext context)
-                }
-
-                DiagnosticDescriptor rule = RuleDoNotUseInsecureDtdProcessingInApiDesign;
-                context.ReportDiagnostic(
-                    CreateDiagnostic(
-                        methodSymbol.Locations,
-                        rule,
-                        SecurityDiagnosticHelpers.GetLocalizableResourceString(
-                            nameof(MicrosoftNetFrameworkAnalyzersResources.XmlTextReaderDerivedClassConstructorNoSecureSettingsMessage),
-                            SecurityDiagnosticHelpers.GetNonEmptyParentName(node, model, context.CancellationToken)
-                        )
-                    )
-                );
+                });
             }
 
-            private void AnalyzeNodeForXmlTextReaderDerivedTypeMethodDecl(SyntaxNodeAnalysisContext context)
+            private void AnalyzeBlockForXmlTextReaderDerivedTypeMethodDecl(
+                OperationBlockStartAnalysisContext context)
             {
-                SyntaxNode node = context.Node;
-                SemanticModel model = context.SemanticModel;
-
-
-                if (SyntaxNodeHelper.GetDeclaredSymbol(node, model) is not IMethodSymbol methodSymbol ||
-                   !((!Equals(methodSymbol.ContainingType, _xmlTypes.XmlTextReader)) && methodSymbol.ContainingType.DerivesFrom(_xmlTypes.XmlTextReader, baseTypesOnly: true)))
+                if (context.OwningSymbol is not IMethodSymbol methodSymbol
+                    || Equals(methodSymbol.ContainingType, _xmlTypes.XmlTextReader)
+                    || !methodSymbol.ContainingType.DerivesFrom(_xmlTypes.XmlTextReader, baseTypesOnly: true))
                 {
                     return;
                 }
 
-                // If the default value are not secure, the AnalyzeNodeForXmlTextReaderDerivedTypeConstructorDecl would be skipped,
-                // therefoer we need to check constructor for any insecure settings.
+                // If the default value are not secure, the AnalyzeNodeForXmlTextReaderDerivedTypeConstructorDecl
+                // would be skipped, therefore we need to check constructor for any insecure settings.
                 // Otherwise, we skip checking constructors
                 if (_isFrameworkSecure && methodSymbol.MethodKind == MethodKind.Constructor)
                 {
                     return;
                 }
 
-                bool hasSetXmlResolver = false;
-                bool hasSetInsecureXmlResolver = true;
-                bool isDtdProcessingSet = false;
-                bool isDtdProcessingEnabled = true;
+                var hasSetXmlResolver = new ThreadSafeBoolean(false);
+                var hasSetInsecureXmlResolver = new ThreadSafeBoolean(true);
+                var isDtdProcessingSet = new ThreadSafeBoolean(false);
+                var isDtdProcessingEnabled = new ThreadSafeBoolean(true);
 
-                List<Location> locs = null;
+                var locations = new ConcurrentQueue<Location>();
```

Comment:
Feel free to suggest a better concurrent collection. I want it to be thread safe and respect the order.

Reply:
We should not rely on order of operation callbacks. Instead, we can probably use a SortedSet which is order by location start.

Reply:
As far as I know `SortedSet` isn't thread-safe so I am going on @dotpaul direction